zendesk/pkg/adapter/zendesk.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright (c) 2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"go.uber.org/zap"
+)
+
+// ZendeskAPIHandler listen for Zendesk API Events
+type ZendeskAPIHandler interface {
+	Start(stopCh <-chan struct{}) error
+}
+
+// constats for the CE data
+const (
+	ceID      = "wrapper.EventID"
+	ceType    = "com.zendesk.ticket.new"
+	ceSource  = "com.zendesk.source"
+	ceSubject = "New Zendesk Ticket"
+)
+
+const (
+	// Response for successfully receiving an event from Zendesk
+	rOK = `{"200" : "Thanks Zendesk!"}`
+	// Response for failing authentication (sometimes used as a prefix to the reason)
+	rAuthFailed = `Authentication FAILED`
+)
+
+type zendeskAPIHandler struct {
+	port     int
+	token    string
+	username string
+	password string
+
+	ceClient cloudevents.Client
+	srv      *http.Server
+
+	logger *zap.SugaredLogger
+}
+
+// NewZendeskAPIHandler creates the default implementation of the Zendesk API Events handler
+func NewZendeskAPIHandler(ceClient cloudevents.Client, port int, token, username, password string, logger *zap.SugaredLogger) ZendeskAPIHandler {
+	return &zendeskAPIHandler{
+		port:     port,
+		token:    token,
+		username: username,
+		password: password,
+
+		ceClient: ceClient,
+		logger:   logger,
+	}
+}
+
+// Start the server for receiving Zendesk events. Will block until the stop channel closes.
+func (h *zendeskAPIHandler) Start(stopCh <-chan struct{}) error {
+	h.logger.Info("Starting Zendesk event handler..")
+
+	m := http.NewServeMux()
+	m.HandleFunc("/", h.handleAll)
+
+	h.srv = &http.Server{
+		Addr:    ":" + strconv.Itoa(h.port),
+		Handler: m,
+	}
+
+	done := make(chan bool, 1)
+	go h.gracefulShutdown(stopCh, done)
+
+	h.logger.Infof("Zendesk Source is ready to handle requests at %s", h.srv.Addr)
+	if err := h.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		return fmt.Errorf("could not listen on %s: %v", h.srv.Addr, err)
+	}
+
+	<-done
+	h.logger.Infof("Server stopped")
+	return nil
+}
+
+func (h *zendeskAPIHandler) authenticate(r *http.Request) (bool, error) {
```

Comment:
I think we could do a different signature:
```suggestion
func (h *zendeskAPIHandler) validateAuthHeader(r *http.Request) error {
```

then the caller checks for error 
```golang
if err := h.validateAuthHeader(req); err != nil {
   return err
}
```